modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C

```diff
@@ -847,3 +841,273 @@ AdvectiveFluxCalculatorBase::PQPlusMinus(dof_id_type sequential_i,
 
   return result;
 }
+
+void
+AdvectiveFluxCalculatorBase::buildCommLists()
+{
+  /**
+   * Build the multi-processor communication lists.
+   *
+   * (A) We will have to send _u_nodal information to other processors.
+   * This is because although we can Evaluate Variables at all elements in
+   * _fe_problem.getEvaluableElementRange(), in the PorousFlow setting
+   * _u_nodal could depend on Material Properties within the elements, and
+   * we can't access those Properties within the ghosted elements.
+   * We don't know which processors require _u_nodal from our nodes.
+   * However, we do know which nodes we require information about, and the
+   * processors that can give us that information.  So we store that in
+   * _nodes_to_receive.  Then we advertise to those processors that we need
+   * the information.  Similarly, we will receive advertisements from the
+   * other processors saying they need information from us.  We then store
+   * that in _nodes_to_send.
+   *
+   * (B) We need to send _kij information to other processors.
+   * The same strategy is followed, first building pairs_to_receive and then
+   * _pairs_to_send.
+   */
+
+  // We use this (hopefully unique) message tag for all communication in this method.
+  // QUESTION: If there are multiple AdvectiveFluxCalculatorBase in the input file, will this
+  // work?
+  Parallel::MessageTag send_tag = _communicator.get_unique_tag(4712);
+
+  _nodes_to_receive.assign(_app.n_processors(), std::vector<dof_id_type>());
+  for (const auto & elem : _fe_problem.getEvaluableElementRange())
+    if (this->hasBlocks(elem->subdomain_id()))
+    {
+      const processor_id_type elem_pid = elem->processor_id();
+      if (elem_pid != _my_pid)
+        for (unsigned i = 0; i < elem->n_nodes(); ++i)
+          if (std::find(_nodes_to_receive[elem_pid].begin(),
+                        _nodes_to_receive[elem_pid].end(),
+                        elem->node_id(i)) == _nodes_to_receive[elem_pid].end())
+            _nodes_to_receive[elem_pid].push_back(elem->node_id(i));
+    }
+
+  // Now advertise that we need this information from the other processors
+  std::vector<Parallel::Request> send_requests(_app.n_processors() - 1);
+  unsigned sr = 0;
+  for (processor_id_type pid = 0; pid < _app.n_processors(); ++pid)
+    if (pid != _my_pid)
+      // non-blocking send
+      _communicator.send(pid, _nodes_to_receive[pid], send_requests[sr++], send_tag);
+
+  // Now receive the adverts from all other processors
+  // QUESTION: what is item_type?  Is there a better way of setting it?
+  const auto item_type = libMesh::Parallel::StandardType<dof_id_type>(&(_nodes_to_receive[0][0]));
```

Comment:
As far as I understand item type tells MPI about the type layout of the data you are trying to communicate. In a heterogeneous environment you could probably have MPI transfer everything as byte buffers. iten_type will make MPI aware of the actual types and will make it take care of little/big endian conversions for example.